engines/buried/sound.cpp

```diff
@@ -0,0 +1,761 @@
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * Additional copyright for this file:
+ * Copyright (C) 1995 Presto Studios, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "audio/audiostream.h"
+#include "audio/mixer.h"
+#include "audio/decoders/wave.h"
+#include "common/archive.h"
+#include "common/events.h"
+#include "common/system.h"
+
+#include "buried/buried.h"
+#include "buried/graphics.h"
+#include "buried/resources.h"
+#include "buried/sound.h"
+
+namespace Buried {
+
+#define TIMED_EFFECT_NONE   0x00
+#define TIMED_EFFECT_VOLUME 0x01
+
+#define SOUND_FLAG_DESTROY_AFTER_COMPLETION 0x01
+
+static inline int clipVolume(int volume) {
+	return CLIP<int>(volume, 0, Audio::Mixer::kMaxChannelVolume);
+}
+
+SoundManager::SoundManager(BuriedEngine *vm) : _vm(vm) {
+	_fileIDFootsteps = -1;
+	_lastAmbient = 1;
+	startup();
+}
+
+SoundManager::~SoundManager() {
+	for (int i = 0; i < kMaxSounds; i++)
+		delete _soundData[i];
+}
+
+bool SoundManager::startup() {
+	_paused = false;
+
+	for (int i = 0; i < kMaxSounds; i++)
+		_soundData[i] = new Sound();
+
+	return true;
+}
+
+void SoundManager::shutDown() {
+	if (_paused)
+		return;
+
+	for (int i = 0; i < kMaxSounds; i++) {
+		delete _soundData[i];
+		_soundData[i] = new Sound();
+	}
+}
+
+void SoundManager::pause(bool shouldPause) {
+	for (int i = 0; i < kMaxSounds; i++)
+		_soundData[i]->pause(shouldPause);
+}
+
+bool SoundManager::setAmbientSound(const Common::String &fileName, bool fade, byte finalVolumeLevel) {
+	// Determine which of the two ambient tracks to use
+	int newAmbientTrack = (_lastAmbient == 0) ? 1 : 0;
+
+	// If this ambient track is currently in use, stop and kill it now
+	if (_soundData[kAmbientIndexBase + newAmbientTrack]->isPlaying()) {
+		delete _soundData[kAmbientIndexBase + newAmbientTrack];
+		_soundData[kAmbientIndexBase + newAmbientTrack] = new Sound();
+	}
+
+	bool retVal = true;
+
+	if (fileName.empty()) {
+		if (fade) {
+			// Set parameters for the current ambient, if there is one
+			if (_soundData[kAmbientIndexBase + _lastAmbient]->isPlaying()) {
+				_soundData[kAmbientIndexBase + _lastAmbient]->_loop = true;
```

Comment:
Occurances of _soundData[kAmbientTrackBase + _lastAmbient/newAmbientTrack] are repeated so frequently, it would likely be cleaner to simply set up reference objects to the old/new array entry, and use that instead.